messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/deployment/injection/InjectedJMSContext.java

```diff
@@ -141,4 +154,25 @@ private Object lookup(String name) {
             }
         }
     }
-}
+
+    /**
+     * Synchronization task, which executes "cleanup" on JMSContext in AfterCompletion phase. BeforeCompletion does nothing.
+     */
+    private class AfterCompletionSynchronization implements Synchronization {
+        protected final Runnable cleanUpMethod;
```

Comment:
Why are you using a Runnable to reference the `AbstractJMSContex#cleanUp` method instead of simply calling context.cleanUp() in `afterCompletion`? 
I don't see the benefit of creating a lambda in that case...

Reply:
Unfrotunatelly in afterCompletion JmsContext is not available - I assume that it is caused by its lifecycle (transaction scope). I've tried it several times - in beforeCompletion it is working, afterCompletion (when transaction is already commited or rolled back) no. This lambda approach seemed like usable solution.

Reply:
According to https://docs.oracle.com/javaee/7/api/javax/transaction/TransactionScoped.html: 
>Any Synchronization.afterCompletion methods will be invoked in an undefined context.

So it might be possible that with Narayana, the transaction context is closed before `afterCompletion` is invoked.

In that case, you could use a lambda or an inner class (to capture the context instance).

Reply:
after some consideration, your remark highlights an issue with the code.

If the JmsContext is not available in the afterCompletion, that means that its underlying contexts have been cleaned up beforehand. So it does not change the current behaviour, right?

It seems that we need to distinguish the cleaning up of the requestScoped and transactionScoped JmsContext.
The requestScoped will have to get a @PreDestroy method to clean up its contexts.
The transactionScoped will not have a @PreDestroy method and instead will register a after completion synchronization.

Does that make sense?

Reply:
yes, it makes sense. 

I've tried with this change to achieve following behavior:
The requestScoped are cleaned up in @predestroy method 
The transactionScoped are cleaned up in afterCompletion listener and also in @predestroy method (which does nothing in case that it was cleaned up before). I know that doubled clean up for the transactionScoped is not necessary, but with almost none performance cost, we have more certainty of cleanup execution.